Async Binding Support (#155)
modesttree/Zenject: Documentation/Async.md

## <a id="async-bindings"></a>Async Extensions <smal><sub>*Experimental*</sub>

## Table Of Contents

* Introduction

>     * <a href="#async-and-di">Async in DI</a>

Reviewer: ```suggestion
    * <a href="#async-and-di">`async` in DI</a>
```

    * <a href="#example">Example</a>
* Advanced
    * <a href="#static-memory-pool">Static Memory Pools</a>
    * <a href="#usingstatement">Using statements and dispose pattern</a>

## Introduction
### <a id="async-and-di"></a>Async in DI

> In dependency injection, the injector resolves dependencies of the target class only once, often after class is first created. In other words, injection is a one time process that does not track the injected dependencies to update them later on. If a dependency is not ready at the moment of injection; either binding would not resolve in case of optional bindings or fail completely throwing an error.

Reviewer: ```suggestion
In dependency injection, the injector resolves dependencies of the target class only once, often after class is first created. In other words, injection is a one time process that does not track the injected dependencies to update them later on. If a dependency is not ready at the moment of injection, either the binding wouldn't resolve in case of optional bindings or would fail completely throwing an error.
```

Author: Applied documentation all changes.

> This is creates a dilemma when implementing dependencies that are resolved asyncroniously. You can design around the DI limitations by carefully architecting your code so that the injection happens after async process is completed. This requires careful planning, increased complexity in setup. It is also prone to errors.

Reviewer: ```suggestion
This creates a dilemma while implementing dependencies that are resolved asynchronous. You can design around the DI limitations by carefully designing your code so that the injection happens after the `async` process is completed. This requires careful planning, which leads to an increased complexity in the setup, and is also prone to errors.
```

> Alternatively you can inject a intermediary object that tracks the result of the async operation. When you need to access the dependency, you can use this intermediary object to check if async task is completed and get the resulting object. With the experimental async support, we would like to provide ways to tackle this problem in Extenject. You can find Async extensions in **Plugins/Zenject/OptionalExtras/Async** folder.

Reviewer: ```suggestion
Alternatively you can inject an intermediary object that tracks the result of the `async` operation. When you need to access the dependency, you can use this intermediary object to check if the `async` task is completed and get the resulting object. With the experimental `async` support, we would like to provide ways to tackle this problem in Extenject. You can find the `async` extensions in the folder **Plugins/Zenject/OptionalExtras/Async**.
```

### <a id="example"></a>Example

> Lets see how we can inject async dependencies through an intermediary object. Async extensions implements `AsyncInject<T>` as this intermediary. You can use it as follows. 

Reviewer: ```suggestion
Lets see how we can inject `async` dependencies through an intermediary object. `async` extensions implement `AsyncInject<T>` as this intermediary. You can use it as follows. 
```

```csharp
public class TestInstaller : MonoInstaller<TestInstaller>
{
    public override void InstallBindings()
    {
         Container.BindAsync<IFoo>().FromMethod(async () =>
            {
                await Task.Delay(100);
                return (IFoo) new Foo();
            }).AsCached();
    }
}

public class Bar : IInitializable, IDisposable
{
    readonly AsyncInject<IFoo> _asyncFoo;
    IFoo _foo;
    public Bar(AsyncInject<IFoo> asyncFoo)
    {
        _asyncFoo = asyncFoo;
    }

    public void Initialize()
    {
        if (!asycFoo.TryGetResult(out _foo))
        {
            asycFoo.Completed += OnFooReady;
        }
    }
       
    private void OnFooReady(IFoo foo)
    {
        _foo = foo;
    }

    public void Dispose()
    {
        asycFoo.Completed -= OnFooReady;
    }

> }

Reviewer: In this class, the `AsyncInject<T>` gets stored in `_asyncFoo`, but `asycFoo` is used later on in the methods, even though this identifier (with additional `n`) is only used for the constructor parameter. It looks like either of them (the `asyncFoo` usages or `_asyncFoo`) need to be adjusted.

Suggestion A: rename member variable to `asyncFoo`:
```suggestion
public class Bar : IInitializable, IDisposable
{
    readonly AsyncInject<IFoo> asyncFoo;
    IFoo _foo;
    public Bar(AsyncInject<IFoo> asyncFoo)
    {
        this.asyncFoo = asyncFoo;
    }

    public void Initialize()
    {
        if (!asyncFoo.TryGetResult(out _foo))
        {
            asyncFoo.Completed += OnFooReady;
        }
    }
       
    private void OnFooReady(IFoo foo)
    {
        _foo = foo;
    }

    public void Dispose()
    {
        asyncFoo.Completed -= OnFooReady;
    }
}
```

Suggestion B: use the current name of the member variable:
```suggestion
public class Bar : IInitializable, IDisposable
{
    readonly AsyncInject<IFoo> _asyncFoo;
    IFoo _foo;
    public Bar(AsyncInject<IFoo> asyncFoo)
    {
        _asyncFoo = asyncFoo;
    }

    public void Initialize()
    {
        if (!_asyncFoo .TryGetResult(out _foo))
        {
            _asyncFoo .Completed += OnFooReady;
        }
    }
       
    private void OnFooReady(IFoo foo)
    {
        _foo = foo;
    }

    public void Dispose()
    {
        _asyncFoo .Completed -= OnFooReady;
    }
}
```
Keep in mind that there is another code snippet using `asycFoo` as well.

Author: I have opted for renaming it to _asyncFoo. I don't personally use _ prefix on my own work. But Extenject code examples use it.

Commited the changes

```

> Here we use `BindAsync<IFoo>().FromMethod()` to pass an async method delegate that waits for 100 ms and then returns a newly created `Foo` object. This method can be any other method with `async Task<T> Method()` signature. `BindAsync<T>` extension provides a separate binder for async operations. This binder is currently limited to a few `FromX()` providers. Features like Pooling and Factories are not supported at the moment.

Reviewer: ```suggestion
Here we use `BindAsync<IFoo>().FromMethod()` to pass an `async` lambda that waits for 100 ms and then returns a newly created `Foo` instance. This method can be any other method with the signature `Task<T> Method()`. *Note: the `async` keyword is an implementation detail and thus not part of the signature.* The `BindAsync<T>` extension method provides a separate binder for `async` operations. This binder is currently limited to a few `FromX()` providers. Features like Pooling and Factories are not supported at the moment.
```

> With above binding `AsyncInject<IFoo>` object is added to the container. Since scope is set to `AsCached()` the operation will run only once and `AsyncInject<IFoo>` will keep the result. It is important to note that async operation will not start before this binding is resolved. If you want async operation to start immediately after installing use `NonLazy()` option. 

Reviewer: ```suggestion
With the above `AsyncInject<IFoo>` binding, the instance is added to the container. Since the scope is set to `AsCached()` the operation will run only once and `AsyncInject<IFoo>` will keep the result. It is important to note that `async` operations wont start before this binding is getting resolved. If you want `async` operation to start immediately after installing, use `NonLazy()`. 
```

> Once injected to `Bar`, we can check whether the return value of the async operation is already available by `TryGetResult`. method. This method will return false if there is no result to return. If result is not ready yet, we can listen to `Completed` event to get the return value when async operation completes.

Reviewer: ```suggestion
Once injected to `Bar`, we can check whether the return value of the `async` operation is already available by calling the `TryGetResult` method. This method will return `false` if there is no result to return. If the result isn't ready yet, we can listen to the `Completed` event to get the return value when the `async` operation completes.
```

> Alternatively we can use following methods to check result.

Reviewer: ```suggestion
Alternatively we can use the following methods to check the results availability.
```
